src/List.h

```diff
@@ -20,67 +20,353 @@
+	{
+	public:
+		using difference_type = std::ptrdiff_t;
+		using size_type = std::size_t;
+		using value_type = T;
+		using pointer = T;
+		using reference = T&;
+		using iterator_category = std::random_access_iterator_tag;
+	};
+}
+
+
+template<typename T>
+class List {
 public:
-	void clear();		// remove all entries
+
+	const int DEFAULT_LIST_SIZE = 10;
+	const int LIST_GROWTH_FACTOR = 2;
+
+	~List()		{ free(entries); }
+	explicit List(int size = 0)
+		{
+		num_entries = 0;
+
+		if ( size <= 0 )
+			{
+			max_entries = 0;
+			entries = nullptr;
+			return;
+			}
+
+		max_entries = size;
+
+		entries = (T*) safe_malloc(max_entries * sizeof(T));
+		}
+
+	List(const List& b)
+		{
+		max_entries = b.max_entries;
+		num_entries = b.num_entries;
+
+		if ( max_entries )
+			entries = (T*) safe_malloc(max_entries * sizeof(T));
+		else
+			entries = nullptr;
+
+		for ( int i = 0; i < num_entries; ++i )
+			entries[i] = b.entries[i];
+		}
+
+	List(List&& b)
+		{
+		entries = b.entries;
+		num_entries = b.num_entries;
+		max_entries = b.max_entries;
+
+		b.entries = nullptr;
+		b.num_entries = b.max_entries = 0;
+		}
+
+	List(const T* arr, int n)
+		{
+		num_entries = max_entries = n;
+		entries = (T*) safe_malloc(max_entries * sizeof(T));
+		memcpy(entries, arr, n * sizeof(T));
+		}
+
+	List(std::initializer_list<T> il) : List(il.begin(), il.size()) {}
+
+	List& operator=(const List& b)
+		{
+		if ( this == &b )
+			return *this;
+
+		free(entries);
+
+		max_entries = b.max_entries;
+		num_entries = b.num_entries;
+
+		if ( max_entries )
+			entries = (T *) safe_malloc(max_entries * sizeof(T));
+		else
+			entries = nullptr;
+
+		for ( int i = 0; i < num_entries; ++i )
+			entries[i] = b.entries[i];
+
+		return *this;
+		}
+
+	List& operator=(List&& b)
+		{
+		if ( this == &b )
+			return *this;
+
+		free(entries);
+		entries = b.entries;
+		num_entries = b.num_entries;
+		max_entries = b.max_entries;
+
+		b.entries = nullptr;
+		b.num_entries = b.max_entries = 0;
+		return *this;
+		}
+
+	// Return nth ent of list (do not remove).
+	T operator[](int i) const
+		{
+#ifdef SAFE_LISTS
+		if ( i < 0 || i > num_entries-1 )
+			return 0;
+		else
+#endif
+			return entries[i];
+		}
+
+	void clear()		// remove all entries
+		{
+		free(entries);
+		entries = nullptr;
+		num_entries = max_entries = 0;
+		}
+
 	int length() const	{ return num_entries; }
 	int max() const		{ return max_entries; }
-	int resize(int = 0);	// 0 => size to fit current number of entries
+	int resize(int new_size = 0)	// 0 => size to fit current number of entries
+		{
+		if ( new_size < num_entries )
+			new_size = num_entries;	// do not lose any entries
 
-	void sort(list_cmp_func cmp_func);
+		if ( new_size != max_entries )
+			{
+			entries = (T*) safe_realloc((void*) entries, sizeof(T) * new_size);
+			if ( entries )
+				max_entries = new_size;
+			else
+				max_entries = 0;
+			}
+
+		return max_entries;
+		}
+
+	void sort(list_cmp_func cmp_func)
```

Comment:
Now that we have iterators, calling `std::sort(begin(this), end(this))` will work fine and afik `std::sort` has better performance than qsort for most things. `std::sort` can also accept a custom comparison operator so maybe we could have an overloaded version that accepts that, and by default just use `T`'s comparison functions.

If we did that we'd still get compile time errors if `sort()` was called using a type without a comparison operator because the template would't be able to be instantiated properly.

Reply:
For `std::sort` there's more things that both the container and iterators need to support for it compile. Namely, the container needs to support `swap()` (implementing `ValueSwappable`) and the iterators need to have support move operations (implementing `MoveAssignable` and `MoveConstructable`). I would love to add support for those, but I think I'd rather push that into the "further work" bucket and get the existing changes merged as a first step.